src/general/graph_coloring.rs

```diff
@@ -0,0 +1,179 @@
+use crate::data_structures::{Graph, UndirectedGraph};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+pub fn color_graph<'a, T>(graph: &'a UndirectedGraph<T>) -> HashMap<&'a T, u32>
+where
+    T: 'a + Eq + Hash + Debug,
+{
+    let empty_vec = Vec::new();
+    graph
+        .nodes()
+        .iter()
+        .fold(HashMap::new(), |mut colors, node| {
+            let used_colors = graph
+                .neighbours(node)
+                .unwrap_or(&empty_vec)
+                .iter()
+                .filter_map(|(neighbor, _)| colors.get(neighbor))
+                .collect::<HashSet<_>>();
+
+            let color = (0..).find(|i| !used_colors.contains(i)).unwrap_or(0);
+            colors.insert(node, color);
+
+            colors
+        })
+}
+
+#[cfg(test)]
+mod test_color_graph {
+    use super::*;
```

Comment:
nit(non-blocking): Could we `use` only what's needed?

Reply:
yeah, you are right. should I create another PR for simplifying `use` here?

Reply:
if you want to yeah 👍